File: 1338. Reduce_Array_Size_to_The_Half/solution.java
class Solution {
    public int minSetSize(int[] arr) {
        
        Map<Integer, Integer> dict = new HashMap<Integer, Integer>();
        int ans = 0;
        
        for(int i = 0 ; i < arr.length ; i++){
            if(dict.containsKey(arr[i])){
                int old_val = dict.get(arr[i]);
                dict.put(arr[i], old_val + 1);
            }
            else{
                dict.put(arr[i], 1);
            }
                
        }
        
        ArrayList<Integer> list = new ArrayList<Integer>();
        
        for(Map.Entry<Integer,Integer> set : dict.entrySet())
            list.add(set.getValue());
        
        Collections.sort(list);
        
        int counter = 0;
        for(int i = list.size() - 1 ; i >= 0 && counter < arr.length / 2 ; i--){
            //System.out.println(list.get(i));
            counter += list.get(i);
            ans++;
        }
        
        return ans;
        
        
    }
}

File: 1338. Reduce_Array_Size_to_The_Half/solution.cpp
class Solution {
public:
    int minSetSize(vector<int>& arr) {
        
        map<int,int> dict;
        vector<int> reverse;
        int ans = 0;
        
        for(int i = 0 ; i < arr.size() ; i++)
            dict[arr[i]]++;
        
        for(auto it = dict.begin() ; it != dict.end() ; it++)
            reverse.push_back(it->second);
        
        sort(reverse.begin(), reverse.end());
        
        int counter = 0;
        for(int i = reverse.size() - 1 ; i >= 0 && counter < arr.size() / 2; i--){
            //cout << counter << " " << reverse[i] << endl;
            counter += reverse[i];
            ans++;
        }
        
        return ans;
        
    }
};